src/irmin-pack/unix/gc.ml

```diff
@@ -17,6 +17,310 @@
 open! Import
 include Gc_intf
 
+module Steps_timer = struct
+  type duration = Stats.Latest_gc.duration = {
+    wall : float;
+    sys : float;
+    user : float;
+  }
+
+  type t = { timer : duration; prev_stepname : string }
+
+  let get_wtime () =
+    (Mtime_clock.now () |> Mtime.to_uint64_ns |> Int64.to_float) /. 1e9
+
+  let get_stime () = Rusage.((get Self).stime)
+  let get_utime () = Rusage.((get Self).utime)
+
+  let create first_stepname =
+    let wall = get_wtime () in
+    let sys = get_stime () in
+    let user = get_utime () in
+    let timer = { wall; sys; user } in
+    { timer; prev_stepname = first_stepname }
+
+  let progress prev next_stepname =
+    let wall = get_wtime () in
+    let sys = get_stime () in
+    let user = get_utime () in
+    let next = { wall; sys; user } in
+
+    let wall = next.wall -. prev.timer.wall in
+    let sys = next.sys -. prev.timer.sys in
+    let user = next.user -. prev.timer.user in
+    let delta = (prev.prev_stepname, { wall; sys; user }) in
+
+    let next = { timer = next; prev_stepname = next_stepname } in
+    (next, delta)
+end
+
+module Main_stats : sig
+  type t
+
+  val create :
+    string ->
+    commit_offset:int63 ->
+    before_suffix_start_offset:int63 ->
+    before_suffix_end_offset:int63 ->
+    after_suffix_start_offset:int63 ->
+    t
+
+  val finish_current_step : t -> string -> t
+
+  val finalise :
+    t ->
+    Stats.Latest_gc.worker ->
+    after_suffix_end_offset:int63 ->
+    Stats.Latest_gc.stats
+end = struct
+  module S = Stats.Latest_gc
+
+  type t = { stats : S.stats; timer : Steps_timer.t }
+  (** [t] is the running state while computing the stats *)
+
+  let create first_stepname ~commit_offset ~before_suffix_start_offset
+      ~before_suffix_end_offset ~after_suffix_start_offset =
+    let stats = Irmin.Type.(random S.stats_t |> unstage) () in
+    (* [repr] provides doesn't provide a generator that fills a type with
+       zeroes but it provides a random generator. Let's use it for our initial
+       value. *)
+    let stats =
+      S.
+        {
+          stats with
+          steps = [];
+          commit_offset;
+          before_suffix_start_offset;
+          before_suffix_end_offset;
+          after_suffix_start_offset;
+        }
+    in
+    let timer = Steps_timer.create first_stepname in
+    { stats; timer }
+
+  let finish_current_step t next_stepname =
+    let timer, prev_step = Steps_timer.progress t.timer next_stepname in
+    let stats = { t.stats with steps = prev_step :: t.stats.steps } in
+    { stats; timer }
+
+  let finalise t worker ~after_suffix_end_offset =
+    let t = finish_current_step t "will not appear in the stats" in
```

Comment:
I'm not sure I understand where you filter this step out

Reply:
`finalise` calls `finish_current_step` to finish the ongoing step. `finish_current_step` takes the name of the next step, there will be no next step